apis/r/R/SOMADenseNdArray.R

```diff
@@ -0,0 +1,144 @@
+    #' element in the array.
+    #' @param shape a vector of integers defining the shape of the array.
+    create = function(type, shape) {
+      stopifnot(
+        "'type' must be a valid Arrow type" =
+          is_arrow_data_type(type),
+        "'shape' must be a vector of positive integers" =
+          is.vector(shape) && all(shape > 0)
+      )
+
+      zstd_filter_list <- tiledb::tiledb_filter_list(c(
+          tiledb_zstd_filter(level = 3L)
+      ))
+
+      # create array dimensions
+      # use tiledb default names like `__dim_0`
+      tdb_dims <- vector(mode = "list", length = length(shape))
+      for (i in seq_along(shape)) {
+        tdb_dims[[i]] <- tiledb::tiledb_dim(
+          name = paste0("soma_dim_", i - 1L),
+          domain = bit64::as.integer64(c(0L, shape[i] - 1L)),
+          tile = bit64::as.integer64(min(c(shape[i], 2048L))),
+          type = "INT64"
+        )
+        tiledb::filter_list(tdb_dims[[i]]) <- zstd_filter_list
+      }
+
+      # create array attribute
+      tdb_attr <- tiledb::tiledb_attr(
+        name = "soma_data",
+        type = tiledb_type_from_arrow_type(type),
+        filter_list = zstd_filter_list
+      )
+
+      # array schema
+      tdb_schema <- tiledb::tiledb_array_schema(
+        domain = tiledb::tiledb_domain(tdb_dims),
+        attrs = tdb_attr,
+        sparse = FALSE,
+        cell_order = "ROW_MAJOR",
+        tile_order = "ROW_MAJOR",
+        capacity=100000,
+        offsets_filter_list = tiledb::tiledb_filter_list(c(
+          tiledb::tiledb_filter("DOUBLE_DELTA"),
+          tiledb::tiledb_filter("BIT_WIDTH_REDUCTION"),
+          tiledb::tiledb_filter("ZSTD")
+        ))
+      )
+
+      # create array
+      tiledb::tiledb_array_create(uri = self$uri, schema = tdb_schema)
+    },
+
+    #' @description Read as an ['arrow::Table']
+    #' @param coords A `list` of integer vectors, one for each dimension, with a
+    #' length equal to the number of values to read. If `NULL`, all values are
+    #' read. List elements can be named when specifying a subset of dimensions.
+    #' @param result_order Order of read results. This can be one of either
+    #' `"ROW_MAJOR, `"COL_MAJOR"`, `"GLOBAL_ORDER"`, or `"UNORDERED"`.
+    #' @return An [`arrow::Table`].
+    read_arrow_table = function(
+      coords = NULL,
+      result_order = "ROW_MAJOR"
+    ) {
+      on.exit(private$close())
+      private$open("READ")
+
+      arr <- self$object
+
+      # select ranges
+      if (!is.null(coords)) {
+        stopifnot(
+          is.list(coords),
+          "'coords' must be a list of vectors" =
+            all(vapply_lgl(coords, is.vector))
+        )
+        tiledb::selected_ranges(arr) <- lapply(coords, function(x) cbind(x, x))
+      }
+
+      # result order
+      tiledb::query_layout(arr) <- match_query_layout(result_order)
+      tiledb::return_as(arr) <- "asis"
+
+      do.call(arrow::arrow_table, arr[])
```

Comment:
It would likely be faster to use the `libtiledbsoma` C++ backing.

Reply:
No doubt. My plan was to use pure R for the initial versions of the foundational classes and then migrate all of them to the c++ reader in a separate PR.

Reply:
Works for me.  We migrate all at once or one by one.  Having tests and established behaviour under them works too.